File: windows/tests/CryptoMako.Vault.Tests/SMBSourceURLTests.cs
using System.Text.Json;
using Xunit;

namespace CryptoMako.Vault.Tests;

public class SMBSourceURLTests
{
    [Fact]
    public void Normalize_basic_share()
    {
        Assert.Equal("smb://files.example/docs", SMBSourceURL.Normalize("smb://files.example/docs"));
    }

    [Fact]
    public void Normalize_with_subpath_and_port()
    {
        Assert.Equal(
            "smb://nas.local:445/backup/photos",
            SMBSourceURL.Normalize("smb://nas.local:445/backup/photos/"));
    }

    [Fact]
    public void Normalize_UNC()
    {
        Assert.Equal("smb://nas/share/folder", SMBSourceURL.Normalize(@"\\nas\share\folder"));
    }

    [Fact]
    public void Normalize_forward_slash_UNC()
    {
        Assert.Equal("smb://nas/share/folder", SMBSourceURL.Normalize("//nas/share/folder"));
    }

    [Fact]
    public void Reject_missing_share()
    {
        var ex = Assert.Throws<SMBSourceURL.ParseException>(() => SMBSourceURL.Normalize("smb://server-only"));
        Assert.Contains("share", ex.Message, StringComparison.OrdinalIgnoreCase);
    }

    [Fact]
    public void Reject_non_SMB()
    {
        Assert.Throws<SMBSourceURL.ParseException>(() => SMBSourceURL.Normalize("https://example/share"));
    }

    [Fact]
    public void ShortLabel_and_ShareName()
    {
        var n = SMBSourceURL.Normalize("smb://nas/media/films");
        Assert.Equal("media", SMBSourceURL.ShareName(n));
        Assert.Equal("nas/media", SMBSourceURL.ShortLabel(n));
    }

    [Fact]
    public void ToUnc_and_ToShareUnc()
    {
        var n = SMBSourceURL.Normalize("smb://nas/media/films");
        Assert.Equal(@"\\nas\media\films", SMBSourceURL.ToUnc(n));
        Assert.Equal(@"\\nas\media", SMBSourceURL.ToShareUnc(n));
    }

    [Fact]
    public void BackupSource_legacy_JSON_defaults_kind_folder()
    {
        var legacy = """{"id":"a","path":"C:\\tmp\\x","vaultFolderName":"X","addedAt":"2020-01-01T00:00:00Z"}""";
        var decoded = JsonSerializer.Deserialize<BackupSource>(legacy, VaultSettings.JsonOptions);
        Assert.NotNull(decoded);
        Assert.Equal(BackupSource.KindFolder, decoded!.Kind);
        Assert.Null(decoded.SmbURL);
        Assert.False(decoded.IsSMB);
    }

    [Fact]
    public void BackupSource_SMB_round_trip_never_stores_password()
    {
        var source = BackupSource.CreateSmb(
            "smb://nas/share",
            @"\\nas\share",
            username: "guille",
            vaultFolderName: "MONSTER");
        Assert.True(source.IsSMB);
        Assert.Equal("smb://nas/share", source.SmbURL);
        Assert.Equal("guille", source.SmbUsername);
        Assert.Equal("smb-password-" + source.Id, source.SmbPasswordAccount);
        Assert.Equal("MONSTER/share", source.VaultFolderName);

        var json = JsonSerializer.Serialize(source, VaultSettings.JsonOptions);
        Assert.DoesNotContain("password", json, StringComparison.OrdinalIgnoreCase);
        Assert.DoesNotContain("guille-secret", json);

        var decoded = JsonSerializer.Deserialize<BackupSource>(json, VaultSettings.JsonOptions);
        Assert.NotNull(decoded);
        Assert.True(decoded!.IsSMB);
        Assert.Equal(source.SmbURL, decoded.SmbURL);
        Assert.Equal("smb://nas/share", source.DisplayLocation);
    }

    [Fact]
    public void SuggestSourcePrefix_UNC_uses_share_name()
    {
        Assert.Equal("docs", BackupSource.SuggestSourcePrefix(@"\\files\docs\photos"));
        Assert.Equal("share", BackupSource.SuggestSourcePrefix(@"\\nas\share"));
    }

    [Fact]
    public void AssertSourceReachable_missing_path_fail_closed()
    {
        var missing = Path.Combine(Path.GetTempPath(), "cryptomako-missing-smb-" + Guid.NewGuid().ToString("N"));
        var ex = Assert.Throws<SMBBackupMount.MountException>(
            () => SMBBackupMount.AssertSourceReachable(missing, isSMB: true));
        Assert.Contains("fail-closed", ex.Message, StringComparison.OrdinalIgnoreCase);
    }

    [Fact]
    public void AssertSourceReachable_missing_UNC_fail_closed()
    {
        var unc = @"\\127.0.0.1\cryptomako-no-such-share-" + Guid.NewGuid().ToString("N");
        var ex = Assert.Throws<SMBBackupMount.MountException>(
            () => SMBBackupMount.AssertSourceReachable(unc, isSMB: true));
        Assert.Contains("fail-closed", ex.Message, StringComparison.OrdinalIgnoreCase);
    }
}

File: Tests/CryptoMakoVaultTests/SMBSourceURLTests.swift
import XCTest
@testable import CryptoMakoShared

final class SMBSourceURLTests: XCTestCase {
    func testNormalizeBasicShare() throws {
        let n = try SMBSourceURL.normalize("smb://files.example/docs")
        XCTAssertEqual(n, "smb://files.example/docs")
    }

    func testNormalizeWithSubpathAndPort() throws {
        let n = try SMBSourceURL.normalize("smb://nas.local:445/backup/photos/")
        XCTAssertEqual(n, "smb://nas.local:445/backup/photos")
    }

    func testNormalizeUNC() throws {
        let n = try SMBSourceURL.normalize(#"\\nas\share\folder"#)
        XCTAssertEqual(n, "smb://nas/share/folder")
    }

    func testRejectMissingShare() {
        XCTAssertThrowsError(try SMBSourceURL.normalize("smb://server-only")) { error in
            XCTAssertEqual(error as? SMBSourceURL.ParseError, .missingShare)
        }
    }

    func testRejectNonSMB() {
        XCTAssertThrowsError(try SMBSourceURL.normalize("https://example/share")) { error in
            XCTAssertEqual(error as? SMBSourceURL.ParseError, .notSMB)
        }
    }

    func testShortLabelAndShareName() throws {
        let n = try SMBSourceURL.normalize("smb://nas/media/films")
        XCTAssertEqual(SMBSourceURL.shareName(from: n), "media")
        XCTAssertEqual(SMBSourceURL.shortLabel(from: n), "nas/media")
    }

    func testBackupSourceCodableDefaultsKindFolder() throws {
        let legacy = """
        {"id":"a","path":"/tmp/x","vaultFolderName":"X","addedAt":0}
        """.data(using: .utf8)!
        let decoded = try JSONDecoder().decode(BackupSource.self, from: legacy)
        XCTAssertEqual(decoded.kind, .folder)
        XCTAssertNil(decoded.smbURL)
        XCTAssertNil(decoded.lastFullSyncAt)
    }

    func testBackupSourceLastFullSyncAtRoundTrip() throws {
        let stamp = Date(timeIntervalSince1970: 1_700_000_000)
        var source = BackupSource(id: "s1", path: "/tmp/y", vaultFolderName: "Y")
        source.lastFullSyncAt = stamp
        let data = try JSONEncoder().encode(source)
        let decoded = try JSONDecoder().decode(BackupSource.self, from: data)
        XCTAssertEqual(decoded.lastFullSyncAt, stamp)
        // Encoded JSON must include the key when set.
        let obj = try JSONSerialization.jsonObject(with: data) as? [String: Any]
        XCTAssertNotNil(obj?["lastFullSyncAt"])
    }

    func testBackupSourceSMBRoundTrip() throws {
        let source = BackupSource(
            id: "smb-1",
            path: "/Volumes/share",
            vaultFolderName: "share",
            kind: .smb,
            smbURL: "smb://nas/share",
            smbUsername: "guille",
            bookmarkData: Data([1, 2, 3])
        )
        let data = try JSONEncoder().encode(source)
        let decoded = try JSONDecoder().decode(BackupSource.self, from: data)
        XCTAssertEqual(decoded, source)
        XCTAssertTrue(decoded.isSMB)
        XCTAssertEqual(decoded.displayLocation, "smb://nas/share")
        XCTAssertEqual(decoded.smbPasswordKeychainAccount, "smb-password-smb-1")
    }

    func testConnectURLIncludesUsernameWithoutPassword() {
        let url = SMBSourceURL.connectURL(normalized: "smb://nas/share/photos", username: "guille")
        XCTAssertEqual(url.scheme, "smb")
        XCTAssertEqual(url.host, "nas")
        XCTAssertEqual(url.user, "guille")
        XCTAssertNil(url.password)
        XCTAssertTrue(url.path.contains("share"))
    }

    func testConnectURLOmitsEmptyUsername() {
        let url = SMBSourceURL.connectURL(normalized: "smb://nas/share", username: "  ")
        XCTAssertNil(url.user)
        XCTAssertEqual(url.absoluteString, "smb://nas/share")
    }
}
